### PR TITLE
More ZIO 2.0 Renames

### LINKED ISSUE #5221: ZIO 2.0 Name Changes
We should rename the old methods to the new names, but also, create the old methods that forward to the new methods, and use the `@deprecated` annotation, marking 2.0.0 as the deprecation version.

 - `ZIO.effect` => `ZIO.attempt`
 - `ZIO.effectTotal` => `ZIO.succeed`
 - `ZIO.effectSuspend` => `ZIO.suspend`
 - `ZIO.effectSuspendTotal` => `ZIO.suspendSucceed`
 - `ZIO.effectSuspendWith` => `ZIO.suspendWith`
 - `ZIO.effectSuspendTotalWith` => `ZIO.suspendSucceedWith`
 - `ZIO#run` => `ZIO#exit`
 - `ZManaged.make` => `ZManaged.bracket`
 - `ZIO#ensuring_` => `ZIO.ensuring`
 - `ZIO#ensuring` => `ZIO.ensuringWith`
 - `ZIO#toManaged_` => `ZIO#toManaged`
 - `ZIO#toManaged` => `ZIO#toManagedWith` / `ZIO#bracketManaged`
 - `putStrLn` => `printLine`
 - `getStrLn` => `readLine`
 
More generally, there is an intention to replace all `_` methods in ZIO 2.0 (`bracket_`, etc.). This naming convention stems from Haskell and has no analogues in the JVM.

### LINKED ISSUE #5224: ZIO 2.0 Forwarders
It would be nice to have forwarders for common functionality, like we do for `ZIO.sleep`.

 - `ZIO.succeedBlocking`
 - `ZIO.attemptBlocking`

### PR DESCRIPTION
**Pull request by adamgfraser**
Resolves #5221. Resolves #5224.

Follow up work on ZIO 2.0 renames. To keep track of things here is a table listing the old and new names of operators. Operating marked as `Done` are already changed in the `series/2.x` brach. Operators marked as `Pending` are done in this PR. Other ones still need to be done and some have proposed changes for discussion.

They fall into a couple of categories:

* `_` suffix variants - we have a lot of these where the return value isn't necessary. In other cases we have used the `XYZ` versus `XYZWith` convention but I don't think that really works with a lot of these. For example, neither `collectAll` nor `collectAll_` take an additional argument beyond the collection so calling one of them `collectAllWith` seems a little weird. We could use a new convention for this.
* `effect` constructors - with the existing changes we are moving away from the `effectXYZ` convention and I think need to do the same thing for other operators. Proposals are below.
* `ZManaged` constructors - there are a bunch of `ZManaged` constructors that fit a little uneasily with the renaming of `make` and `makeExit` to `bracket` and `bracketExit`.

Also, I am thinking we should change all `fromEffect` operators (e.g. `ZManaged.fromEffect`) to `fromZIO`.

@jdegoes Can you let me know if you are on board with these changes and if you have any ideas on the question marks?

Data Type | Old Operator Name | New Operator Name | Status 
-- | -- | -- | --
Chunk | collectM | collectZIO |
Chunk | collectWhileM | collectWhileZIO |
Chunk | dropWhileM | dropWhileZIO |
Chunk | filterM | filterZIO |
Chunk | findM | findZIO |
Chunk | foldM | foldZIO |
Chunk | foldWhileM | foldWhileZIO |
Chunk | mapAccumM | mapAccumZIO |
Chunk | mapM | mapZIO |
Chunk | mapM_ | mapZIODiscard |
Chunk | mapMPar | mapZIOPar |
Chunk | mapMPar_ | mapZIOParDiscard |
Chunk | takeWhileM | takeWhileZIO |
Chunk | unfoldM | unfoldZIO |
Exit | flatMapM | flatMapZIO |
Exit | mapM | mapZIO |
Fiber | mapM | mapZIO |
NonEmptyChunk | mapAccumM | mapAccumZIO |
NonEmptyChunk | mapM | mapZIO |
NonEmptyChunk | mapMPar | mapZIOPar |
Runtime | unsafeRunAsync_ | unsafeRunAsync | Pending
Runtime | unsafeRunAsync | unsafeRunAsyncWith | Pending
Schedule | addDelayM | addDelayZIO |
Schedule | checkM | checkZIO |
Schedule | contramapM | contramapZIO |
Schedule | delayedM | delayedZIO |
Schedule | dimapM | dimapZIO |
Schedule | foldM | foldZIO |
Schedule | mapM | mapZIO |
Schedule | modifyDelayM | modifyDelayZIO |
Schedule | reconsiderM | reconsiderZIO |
Schedule | untilInputM | untilInputZIO |
Schedule | untilOutputM | untilOutputZIO |
Schedule | whileInputM | whileInputZIO |
Schedule | whileOutputM | whileOutputZIO |
Schedule | collectWhileM | collectWhileZIO |
Schedule | collectUntilM | collectUntilZIO |
Schedule | recurWhileM | recurWhileZIO |
Schedule | recurUntilM | recurUntilZIO |
ZHub | contramapM | contramapZIO |
ZHub | dimapM | dimapZIO |
ZHub | filterInputM | filterInputZIO |
ZHub | filterOuputM | filterOutputZIO |
ZHub | mapM | mapZIO |
ZIO | bracket | acquireReleaseWith |
ZIO | bracket_ | acquireRelease |
ZIO | collectM | collectZIO |
ZIO | filterOrElse | filterOrElseWith |
ZIO | filterOrElse_ | filterOrElse |
ZIO | foldCauseM | foldCauseZIO |
ZIO | foldM | foldZIO |
ZIO | foldTraceM | foldTraceZIO |
ZIO | onExit | ensuringWith |
ZIO | rejectM | rejectZIO |
ZIO | repeatUntilM | repeatUntilZIO |
ZIO | repeatWhileM | repeatWhileZIO |
ZIO | retryUntilM | retryUntilZIO |
ZIO | retryWhileM | retryWhileZIO |
ZIO | someOrElseM | someOrElseZIO |
ZIO | unlessM | unlessZIO |
ZIO | whenM | whenZIO |
ZIO | accessM | accessZIO |
ZIO | collectAll_ | collectAllDiscard |
ZIO | collectAllPar_ | collectAllParDiscard |
ZIO | collectAllParN_ | collectAllParNDiscard |
ZIO | effect | attempt | Done
ZIO | effectAsync | async |
ZIO | effectAsyncInterrupt | asyncInterrupt 
ZIO | effectAsyncM | effectAsyncZIO |
ZIO | effectAsyncMaybe | asyncMaybe |
ZIO | effectBlocking | attemptBlocking | Pending
ZIO | effectBlockingCancelable | attemptBlockingCancelable |
ZIO | effectBlockingIO | attemptBlockingIO |
ZIO | effectBlockingInterrupt | attemptBlockingInterrupt |
ZIO | effectSuspend | suspend | Done
ZIO | effectSuspendMaybeWith | suspendMaybeWith |
ZIO | effectSuspendTotal | suspendSucceed | Done
ZIO | effectSuspendTotalWith | suspendSucceedWith | Done
ZIO | effectSuspendWith | suspendWith | Done
ZIO | effectTotal | succeed | Done
ZIO | foreach_ | foreachDiscard |
ZIO | foreachPar_ | foreachParDiscard |
ZIO | foreachParN_ | foreachParNDiscard |
ZIO | forkAll_ | forkAllDiscard |
ZIO | fromFiberM | fromFiberZIO |
ZIO | fromFunctionM | fromFunctionZIO |
ZIO | ifM | ifZIO |
ZIO | loop_ | loopDiscard |
ZIO | replicateM | replicateZIO |
ZIO | replicateM_ | replicateZIODiscard |
ZIO | unlessM | unlessZIO |
ZIO | whenCaseM | whenCaseZIO |
ZIO | whenM | whenZIO |
ZIO | validate_ | validateDiscard |
ZIO | validatePar_ | validateParDiscard |
ZLayer | fromFunctionM | fromFunctionZIO |
ZLayer | fromFunctionManyM | fromFunctionManyZIO |
ZLayer | fromServiceM | fromServiceZIO |
ZLayer | fromServicesM | fromServicesZIO |
ZLayer | fromServiceManyM | fromServiceManyZIO |
ZLayer | fromServicesManyM | fromServicesManyZIO |
ZManaged | mapM | mapZIO |
ZManaged | onExit | ensuringWith |
ZManaged | onExitFirst | ensuringWithFirst |
ZManaged | use_ | useDiscard |
ZManaged | collectAll_ | collectAllDicard |
ZManaged | collectAllPar_ | collectAllParDiscard |
ZManaged | collectAllParN_ | collecrAllParNDiscard |
ZManaged | foreach_ | foreachDiscard |
ZManaged | foreachPar_ | foreachParDiscard |
ZManaged | foreachParN_ | foreachParNDiscard |
ZManaged | loop_ | loopDiscard |
ZManaged | make | acquireReleaseWith |
ZManaged | make_ | acquireRelease |
ZManaged | makeEffect | acquireReleaseAttemptWith |
ZManaged | makeEffect_ | acquireReleaseAttempt |
ZManaged | makeEffectTotal | acquireReleaseSucceedWith |
ZManaged | makeEffectTotal_ | acquireReleaseSucceed |
ZManaged | makeExit | acquireReleaseExitWith |
ZManaged | makeExit_ | acquireReleaseExit |
ZManaged | makeInterruptible | acquireReleaseInterruptibleWith |
ZManaged | makeInterruptible_ | acquireReleaseInterruptible |
ZManaged | makeReserve | fromReservation |
ZQueue | contramapM | contrmapZIO |
ZQueue | dimapM | dimapZIO |
ZQueue | filterInputM | filterInputZIO |
ZQueue | filterOutputM | filterOutputZIO |
ZQueue | mapM | mapZIO |
ZRef | collectM | collectZIO |
ZRef | contramapM | contrmapZIO |
ZRef | dimapM | dimapZIO |
ZRef | filterInputM | filterInputZIO |
ZRef | filterOutputM | filterOutputZIO |
ZRef | foldAllM | foldAllZIO |
ZRef | foldM | foldZIO
ZRef | mapM | mapZIO |
ZRef | getAndUpdateM | getAndUpdateZIO |
ZRef | getAndUpdateSomeM | getAndUpdateSomeZIO |
ZRef | modifyM | modifyZIO |
ZRef | modifySomeM | modifySomeZIO |
ZRef | updateM | updateZIO |
ZRef | updateAndGetM | updateAndGetZIO |
ZRef | updateSomeM | updateSomeZIO |
ZRef | updateSomeAndGetM | updateSomeAndGetZIO |
ZSTM | collectM | collectSTM |
ZSTM | filterOrElse_ | filterOrElse |
ZSTM | filterOrElse | filterOrElseWith |
ZSTM | foldM | foldSTM |
ZSTM | rejectM | rejectSTM |
ZSTM | accessM | accessSTM |
ZSTM | foreach_ | foreachDiscard |
ZSTM | fromFunctionM | fromFunctionSTM |
ZSTM | ifM | ifSTM |
ZSTM | loop_ | loopDiscard |
ZSTM | replicateM | replicateSTM |
ZSTM | replicateM_ | replicateSTMDiscard |
ZSTM | unlessM | unlessSTM |
ZSTM | whenM | whenSTM |
ZSTM | whenCaseM | whenCaseSTM |
TArray | collectFirstM | collectFirstSTM |
TArray | countM | countSTM |
TArray | existsM | existsSTM |
TArray | findLastM | findLastSTM |
TArray | findM | findSTM |
TArray | foldM | foldSTM |
TArray | forallM | forallSTM |
TArray | indexWhereM | indexWhereSTM |
TArray | reduceOptionM | reduceSTM |
TArray | transformM | transformSTM |
TArray | updateM | updateSTM |
TMap | foldM | foldSTM |
TMap | transformM | transformSTM |
TMap | transformValuesM | transformValuesSTM |
TSet | foldM | foldSTM |
TSet | transformM | transformSTM |
ZTRef | foldAllM | foldAllSTM |
ZTRef | foldM | foldSTM |
Take | foldM | foldZIO |
ZChannel | mapM | mapZIO |
ZPipeline | mapM | mapZIO |
ZSink | contramapM | contramapZIO |
ZSink | contramapChunksM | contramapChunksZIO |
ZSink | dimapM | dimapZIO |
ZSink | dimapChunksM | dimapChunksZIO |
ZSink | foldChunksM | foldChunksZIO |
ZSink | foldM | foldZIO |
ZSink | mapM | mapZIO |
ZSink | untilOutputM | untilOutputZIO |
ZSink | foldLeftM | foldLeftZIO |
ZSink | foldLeftChunksM | foldLeftChunksZIO |
ZStream | mapM | mapZIO |
ZTransducer | mapM | mapZIO |
BoolAlgebra | mapM | mapZIO |
Gen | mapM | mapZIO |
TestAspect | around | aroundWith |
TestAspect | around_ | around |
TestAspect | aroundAll | aroundAllWith
TestAspect | aroundAll_ | aroundAll

### CODE REVIEW

**Comment by iravid** [comment]
Suggestion for the `foreach_` variants - `foreachDiscard`. Agree that using `foreachWith` for the current `foreach` would be awkward. The usage of the `_` variants is an optimization after all, as you do fine with `foreach` and ignoring its result; its usage would therefore be less common.

**Comment by adamgfraser** [comment]
Yes. Could also do `foreachUnit` along similar lines.

**Comment by jdegoes** [comment]
Just brainstorming on convention here:

- A `with` variant could be used if the operator accepts a function to perform the operation "with". All of these cases should be parameterized operators.
- A "with-less" variant could be used if the operator does not accept a function, and returns useful information.
- A "discard" variant could be used if the _callee_ returns no useful information.

**Comment by adamgfraser** [comment]
I think that generally makes sense. One potential issue to think about there is `bracket` and `bracket_` since I think that implies `bracket` should be renamed to `bracketWith` which is bit unfortunate for such a fundamental operator.

**Comment by adamgfraser** [comment]
Updated the table with the above suggestions if people want to see how everything looks. Seems pretty good other than the `bracket` / `ZManaged` variants. For the discard variants could also consider `AndDiscard`, e.g. `collectAllParAndDiscard`, which is wordy but maybe is a little more like English than a collection of keywords.

**Comment by iravid** [comment]
Re bracket - if we’re going from first principles here then bracket is a totally arbitrary word that’s not directly associated with resources. We could name it tryAcquireWith, for example.

**Comment by adamgfraser** [comment]
I like that idea. Maybe `bracket` goes to `acquireReleaseWith` and `bracket_` goes to `acquireRelease`?

**Comment by iravid** [comment]
I like `acquireRelease` 👍🏻 I suggested tryAcquire because that refers to Java's try-with-resources, which has been a helpful analogy in the past, but not sure how important that is.

**Comment by jdegoes** [comment]
@adamgfraser I like that.

**Comment by adamgfraser** [comment]
@jdegoes Okay updated table with proposals for everything. Can you confirm you are good with the above and then I will make the changes?

**Comment by jdegoes** [comment]
@adamgfraser This looks mostly perfect.

However, thinking about what we want to do with: `fromEffect`. It is a convention in some libraries to call such an operator "evaluate" or "execute", e.g. `ZStream.eval(putStrLn("Hello"))`. On the one hand, it breaks with the `fromXYZ` convention that is common in ZIO libraries for alternate (non-primary) constructors that do data type conversion. On the other hand, it somewhat reflects intention: the intention to embed execution of a ZIO effect inside another effect. 

The reason I bring this up now is because it has implications for `mapM`. If `ZStream.fromEffect` were renamed to `ZStream.eval`, then this would imply `mapM` should be renamed to `mapEval`. Similarly for all `M` variants.

The `M`, like the `_`, comes from Haskell and does not have an easy explanation.

**Comment by adamgfraser** [comment]
@jdegoes Yeah, I agree about the `M`. I'm not sure about `Eval` though because that potentially requires a significant amount of explanation of you want to evaluate it but you don't want to evaluate it now you actually want to evaluate it in the context of this other structure. What about `mapZIO`?

**Comment by iravid** [comment]
+1 for eval

**Comment by adamgfraser** [comment]
Updated the list with the `M` suffix operators. One other thing to think about there is what we do with data types that have `M` suffices like `RefM`.

**Comment by jdegoes** [comment]
@adamgfraser `mapZIO`, `fromZIO`, etc. It can work, is close to our existing naming convention (`mapError`, etc.) and it also generalizes to effects beyond `ZIO`, potentially, e.g. `ZIO#mapZPure`, although I don't see that being common.

**Comment by iravid** [comment]
Importing zio.prelude._ could bring in syntax for mapZPure etc; importing zio.interop.cats._ would bring in syntax for mapF. It does generalize better than mapEval.

**Comment by adamgfraser** [comment]
Even in our own code base we use `mapM` for `ZIO` and `STM` right now whereas `mapZIO` and `mapSTM` actually seems quite clear.

**Comment by adamgfraser** [comment]
Okay, updated the table with everything. I think we have a plan here. Only thing I think we have not addressed is `RefM`. I guess in this world it would be `RefZIO` but that feels a little weird.

**Comment by jdegoes** [comment]
@adamgfraser This is looking really good. 😍 

On `RefM`, one idea would be to make it a subtype of `Ref` that is contained in the `Ref` companion object (`Ref.Synchronized`?), and which is constructed with an alternate constructor (`makeSynchronized`). So the question becomes less relevant since it's not a top-level type, just a variation of an existing one.

**Comment by adamgfraser** [comment]
@jdegoes It is a already a subtype of `Ref` contained in the `Ref` companion object. At this point we are just exporting `RefM` as a type alias. So should be an easy change.

**Comment by jdegoes** [comment]
@adamgfraser Superb! Is this ready for final review? Anything else we should consider at the same time?

File: core-tests/shared/src/test/scala/zio/stm/TMapSpec.scala [hunk 1]
```diff
@@ -238,7 +238,7 @@ object TMapSpec extends ZIOBaseSpec {
         val tx =
           for {
             tmap <- TMap.make("a" -> 1, "aa" -> 2, "aaa" -> 3)
-            _    <- tmap.transformM((k, v) => STM.succeed(k.replaceAll("a", "b") -> v * 2))
+            _    <- tmap.transformSTM((k, v) => STM.succeed(k.replaceAll("a", "b") -> v * 2))
```
Comment (jdegoes) [inline core-tests/shared/src/test/scala/zio/stm/TMapSpec.scala:241]: ❤️❤️❤️ This convention is so much nicer than before. We were using `M` to mean different things in different contexts, and that's quite apparent now.

File: core/jvm/src/main/scala/zio/interop/javaz.scala [hunk 2]
```diff
@@ -66,7 +66,7 @@ private[zio] object javaz {
         if (cf.isDone) {
           unwrapDone(p.fatal)(cf)
         } else {
-          Task.effectAsync { cb =>
+          Task.async { cb =>
```
Comment (jdegoes) [inline core/jvm/src/main/scala/zio/interop/javaz.scala:69]: Much prettier!

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 3]
```diff
@@ -821,7 +898,14 @@ sealed abstract class Chunk[+A] extends ChunkLike[A] { self =>
   /**
    * Takes all elements so long as the effectual predicate returns true.
    */
+  @deprecated("use takeWhileZIO", "2.0.0")
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/Chunk.scala:901]: Thank you for all the `@deprecated` warnings! It will help early users of 2.0 release candidates.

**Comment by jdegoes** [comment]
One thing I think we should standardize on:

 - If all the variants of a method end in `XYZ`, then we should omit `XYZ`. e.g. if we have 4 methods `fooXWith`, `fooYWith`, `fooZWith`, then we should instead rename them all `fooX`, `fooY`, `fooZ`, because `with` is not needed for disambiguation.

In words, "disambiguate only when necessary".

**Comment by adamgfraser** [comment]
@jdegoes I think that is a good principle. Is there a place where we are not doing that?

**Comment by adamgfraser** [comment]
In ZIO Test `testM` and `suiteM` need alternatives.

**Comment by adamgfraser** [comment]
The `testM` situation is interesting. In code involving transactions or managed effects a very common pattern is for someone to commit the transaction or use the managed resource so there could be value in having `testZIO`, `testSTM`, and so on, but I don't know if the extra couple of characters to write a test would be acceptable.

**Comment by jdegoes** [comment]
How about a polymorphic `test` that uses implicits to distinguish between a ZIO value and a non-ZIO value? A "smart test" and "smart assert"?

**Comment by adamgfraser** [comment]
Yes I think we could do something there. Let me work on it. May also make sense to do in a follow up PR.

**Comment by jdegoes** [comment]
Agreed to do in followup. Let me know when this one is ready!

**Comment by adamgfraser** [comment]
@jdegoes This is ready.

**Review by jdegoes — APPROVED** [review]

**Comment by jdegoes** [comment]
A colossal undertaking. Prepares us for the next phase of ZIO. Thank you for your work on this! 🙏

**Pull request merged**